### PR TITLE
Iox #27 create server c binding

### PR DESCRIPTION
**Pull request by elfenpiff**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/advanced/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Notes for Reviewer
Implements and tests the server c binding. Please check if
* The C API is complete - every C++ method can be mapped to a free c function
* The functionality is complete. Is it attachable to the WaitSet/Listener? Is there something missing.

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
    - [x] Each unit test case has a unique UUID
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Links to #27

### CODE REVIEW

File: iceoryx_binding_c/test/moduletests/test_server.cpp [hunk 1]
```diff
@@ -0,0 +1,428 @@
...
+    const void* requestPayload;
+    EXPECT_THAT(iox_server_take_request(sut, &requestPayload), Eq(ServerRequestResult_SUCCESS));
+
+    void* payload = nullptr;
+    EXPECT_THAT(iox_server_loan_aligned_response(sut, requestPayload, &payload, sizeof(int64_t), 16),
+                Eq(AllocationResult_SUCCESS));
+    EXPECT_THAT(payload, Ne(nullptr));
+    EXPECT_THAT(reinterpret_cast<uint64_t>(payload) % 16, Eq(0));
+}
+
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_binding_c/test/moduletests/test_server.cpp:385]: Please add a test where `iox_server_loan_aligned_response` fails.

File: iceoryx_posh/source/popo/server_options.cpp [hunk 2]
```diff
@@ -58,5 +58,12 @@ ServerOptions::deserialize(const cxx::Serialization& serialized) noexcept
 
     return cxx::success<ServerOptions>(serverOptions);
 }
+
+bool ServerOptions::operator==(const ServerOptions& rhs) const noexcept
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_posh/source/popo/server_options.cpp:62]: Please update the copyright year.

File: iceoryx_posh/include/iceoryx_posh/popo/server_options.hpp [hunk 3]
```diff
@@ -54,6 +54,10 @@ struct ServerOptions
     /// @brief deserialization of the ServerOptions
     static cxx::expected<ServerOptions, cxx::Serialization::Error>
     deserialize(const cxx::Serialization& serialized) noexcept;
+
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_posh/include/iceoryx_posh/popo/server_options.hpp:57]: Please update the copyright year.

File: iceoryx_hoofs/include/iceoryx_hoofs/error_handling/error_handling.hpp [hunk 4]
```diff
@@ -195,6 +195,7 @@ namespace iox
     error(BINDING_C__C2CPP_ENUM_TRANSLATION_INVALID_CLIENT_STATE_VALUE) \
     error(BINDING_C__C2CPP_ENUM_TRANSLATION_INVALID_SERVER_EVENT_VALUE) \
     error(BINDING_C__C2CPP_ENUM_TRANSLATION_INVALID_SERVER_STATE_VALUE) \
+    error(BINDING_C__C2CPP_ENUM_TRANSLATION_INVALID_SERVER_REQUEST_RESULT_VALUE) \
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/include/iceoryx_hoofs/error_handling/error_handling.hpp:198]: Where is this used?

File: iceoryx_binding_c/test/moduletests/test_listener.cpp [hunk 5]
```diff
@@ -412,16 +428,23 @@ void notifyClient(ClientPortData& portData)
     EXPECT_FALSE(portData.m_chunkReceiverData.m_conditionVariableDataPtr->m_semaphore.post().has_error());
 }
 
+void notifyServer(ServerPortData& portData)
+{
+    iox::popo::ChunkQueuePusher<ServerChunkQueueData_t> pusher{&portData.m_chunkReceiverData};
+    pusher.push(iox::mepoo::SharedChunk());
+    EXPECT_FALSE(portData.m_chunkReceiverData.m_conditionVariableDataPtr->m_semaphore.post().has_error());
+}
+
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_binding_c/test/moduletests/test_listener.cpp:437]: Can you please move this directly above the server tests?

File: iceoryx_binding_c/test/moduletests/test_listener.cpp [hunk 6]
```diff
@@ -447,4 +470,55 @@ TIMING_TEST_F(iox_listener_test, NotifyingClientEventWithContextDataWorks, Repea
...
+    iox_server_t server = iox_server_init(&clientStorage, "ServiceA", "InstanceA", "EventA", nullptr);
+
+    EXPECT_THAT(iox_listener_size(&m_sut), Eq(0));
+    iox_listener_attach_server_event(&m_sut, server, ServerEvent_REQUEST_RECEIVED, &serverCallback);
+    EXPECT_THAT(iox_listener_size(&m_sut), Eq(1));
+
+    iox_listener_detach_server_event(&m_sut, server, ServerEvent_REQUEST_RECEIVED);
+    EXPECT_THAT(iox_listener_size(&m_sut), Eq(0));
+}
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_binding_c/test/moduletests/test_listener.cpp:487]: Please add a `AttachingServerWithContextDataWorks` test.

File: iceoryx_binding_c/test/moduletests/test_wait_set.cpp [hunk 7]
```diff
@@ -139,6 +145,17 @@ void clientCallbackWithContextData(iox::popo::UntypedClient* client, void* const
     iox_ws_test::m_callbackOrigin = client;
     iox_ws_test::m_contextData = contextData;
 }
+
+void serverCallback(iox::popo::UntypedServer* client)
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_binding_c/test/moduletests/test_wait_set.cpp:149]:
```suggestion
void serverCallback(iox::popo::UntypedServer* server)
```

File: iceoryx_binding_c/test/moduletests/test_wait_set.cpp [hunk 8]
```diff
@@ -139,6 +145,17 @@ void clientCallbackWithContextData(iox::popo::UntypedClient* client, void* const
...
+{
+    iox_ws_test::m_callbackOrigin = client;
+}
+
+void serverCallbackWithContextData(iox::popo::UntypedServer* client, void* const contextData)
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_binding_c/test/moduletests/test_wait_set.cpp:154]:
```suggestion
void serverCallbackWithContextData(iox::popo::UntypedServer* server, void* const contextData)
```

File: iceoryx_binding_c/test/moduletests/test_wait_set.cpp [hunk 9]
```diff
@@ -880,4 +891,137 @@ TEST_F(iox_ws_test, NotifyingClientStateWithContextDataWorks)
...
+
+    iox_server_t server = iox_server_init(&serverStorage, "ServiceA", "InstanceA", "EventA", nullptr);
+
+    EXPECT_THAT(iox_ws_size(m_sut), Eq(0));
+    iox_ws_attach_server_event(m_sut, server, ServerEvent_REQUEST_RECEIVED, 0, nullptr);
+    EXPECT_THAT(iox_ws_size(m_sut), Eq(1));
+
+    iox_ws_detach_server_event(m_sut, server, ServerEvent_REQUEST_RECEIVED);
+    EXPECT_THAT(iox_ws_size(m_sut), Eq(0));
+}
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_binding_c/test/moduletests/test_wait_set.cpp:909]: Please add a `AttachingServerEventWithContextDataWorks` test.

File: iceoryx_binding_c/test/moduletests/test_wait_set.cpp [hunk 10]
```diff
@@ -880,4 +891,137 @@ TEST_F(iox_ws_test, NotifyingClientStateWithContextDataWorks)
...
+    EXPECT_THAT(m_callbackOrigin, Eq(static_cast<void*>(server)));
+    EXPECT_THAT(m_contextData, Eq(static_cast<void*>(&serverStorage)));
+
+    iox_ws_detach_server_event(m_sut, server, ServerEvent_REQUEST_RECEIVED);
+}
+
+TEST_F(iox_ws_test, AttachingServerStateWorks)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "7a29c16e-f164-420e-8824-785df24ae8f3");
+    iox_server_storage_t serverStorage;
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_binding_c/test/moduletests/test_wait_set.cpp:968]:
```suggestion
```

File: iceoryx_posh/include/iceoryx_posh/internal/popo/ports/client_server_port_types.hpp [hunk 11]
```diff
@@ -83,7 +83,7 @@ enum class ServerEvent : EventEnumIdentifier
     REQUEST_RECEIVED
 };
 
-enum class ServerState : EventEnumIdentifier
+enum class ServerState : StateEnumIdentifier
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_posh/include/iceoryx_posh/internal/popo/ports/client_server_port_types.hpp:86]: Please update the copyright header. Seems long forgotten...

File: iceoryx_binding_c/test/moduletests/test_wait_set.cpp [hunk 12]
```diff
@@ -880,4 +891,137 @@ TEST_F(iox_ws_test, NotifyingClientStateWithContextDataWorks)
...
+    EXPECT_FALSE(portData.m_chunkReceiverData.m_conditionVariableDataPtr->m_semaphore.post().has_error());
+}
+
+TEST_F(iox_ws_test, NotifyingServerEventWorks)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "003b92c9-3607-4247-8385-3d03085fa574");
+    EXPECT_CALL(*runtimeMock, getMiddlewareServer(_, _, _)).WillOnce(Return(&serverPortData));
+
+    iox_server_t server = iox_server_init(&serverStorage, "ServiceA", "InstanceA", "EventA", nullptr);
+    iox_ws_attach_server_event(m_sut, server, ServerEvent_REQUEST_RECEIVED, 1313799, &serverCallback);
```
Comment (elBoberido) [inline iceoryx_binding_c/test/moduletests/test_wait_set.cpp:924]: magic number `1313799`

File: iceoryx_binding_c/test/moduletests/test_wait_set.cpp [hunk 13]
```diff
@@ -880,4 +891,137 @@ TEST_F(iox_ws_test, NotifyingClientStateWithContextDataWorks)
...
+}
+
+TEST_F(iox_ws_test, NotifyingServerEventWithContextDataWorks)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "a08df13b-ad50-4753-9a1d-07b306d6f8d2");
+    EXPECT_CALL(*runtimeMock, getMiddlewareServer(_, _, _)).WillOnce(Return(&serverPortData));
+
+    iox_server_t server = iox_server_init(&serverStorage, "ServiceA", "InstanceA", "EventA", nullptr);
+    iox_ws_attach_server_event_with_context_data(
+        m_sut, server, ServerEvent_REQUEST_RECEIVED, 1319955, &serverCallbackWithContextData, &serverStorage);
```
Comment (elBoberido) [inline iceoryx_binding_c/test/moduletests/test_wait_set.cpp:948]: again, magic number

File: iceoryx_binding_c/test/moduletests/test_wait_set.cpp [hunk 14]
```diff
@@ -880,4 +891,137 @@ TEST_F(iox_ws_test, NotifyingClientStateWithContextDataWorks)
...
+    EXPECT_THAT(iox_ws_size(m_sut), Eq(0));
+}
+
+TEST_F(iox_ws_test, NotifyingServerStateWorks)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "0fc37898-5aa1-416c-ab19-e64a428ad5ba");
+    EXPECT_CALL(*runtimeMock, getMiddlewareServer(_, _, _)).WillOnce(Return(&serverPortData));
+
+    iox_server_t server = iox_server_init(&serverStorage, "ServiceA", "InstanceA", "EventA", nullptr);
+    iox_ws_attach_server_state(m_sut, server, ServerState_HAS_REQUEST, 51313799, &serverCallback);
```
Comment (elBoberido) [inline iceoryx_binding_c/test/moduletests/test_wait_set.cpp:987]: another one

File: iceoryx_binding_c/test/moduletests/test_server.cpp [hunk 15]
```diff
@@ -0,0 +1,428 @@
...
+    EXPECT_THAT(iox_server_take_request(sut, &payload), Eq(ServerRequestResult_NO_PENDING_REQUESTS));
+}
+
+TEST_F(iox_server_test, WhenOfferedAndRequestsPresentTakeSucceeds)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "42443d18-5a8f-40ee-938d-2b0a345f829b");
+    prepareServerInit();
+    iox_server_t sut = iox_server_init(&sutStorage, SERVICE, INSTANCE, EVENT, nullptr);
+    iox_server_offer(sut);
+    receiveRequest(64461001);
```
Comment (elBoberido) [inline iceoryx_binding_c/test/moduletests/test_server.cpp:230]: Magic number. It seems the parameter for `receiveRequest` is only used in one tests. Could make the parameter optional with a default value and remove the magic numbers from the tests where they aren't checked

File: iceoryx_binding_c/include/iceoryx_binding_c/listener.h [hunk 16]
```diff
@@ -141,4 +142,36 @@ void iox_listener_detach_client_event(iox_listener_t const self,
                                       iox_client_t const client,
                                       const ENUM iox_ClientEvent clientEvent);
 
+/// @brief Attaches a server event to the listener
+/// @param[in] self listener to which the event should be attached to
+/// @param[in] server server which emits the event
```
[the remark this answers is not in the document]
Comment (elfenpiff) [reply iceoryx_binding_c/include/iceoryx_binding_c/listener.h:147]: Yeah it is the new song of Modern Talking, after Cherry Cherry Lady and Brother Louie Louie it is You My Heart you My Server Server from the album Server Server Zero Copy

**Review by FerdinandSpitzschnueffler — APPROVED** [review]

**Review by FerdinandSpitzschnueffler — APPROVED** [review]

**Review by elBoberido — APPROVED** [review]

**Pull request merged**